etcd-defrag/etcddefrag_controller.go

```diff
@@ -0,0 +1,310 @@
+)
+
+const (
+	pollWaitDuration                       = 2 * time.Second
+	pollTimeoutDuration                    = 60 * time.Second
+	maxDefragFailuresBeforeDegrade         = 3
+	minDefragBytes                 int64   = 100 * 1024 * 1024 // 100MB
+	minDefragWaitDuration                  = 36 * time.Second
+	maxFragmentedPercentage        float64 = 45
+
+	controllerRequeueDuration = 10 * time.Minute
+)
+
+type DefragController struct {
+	client.Client
+	log logr.Logger
+
+	ControllerName string
+	HCPNamespace   string
+	upsert.CreateOrUpdateProvider
+
+	etcdClient         etcdcli.EtcdClient
+	numDefragFailures  int
+	requeueTime        time.Time
+	defragWaitDuration time.Duration
+}
+
+func (r *DefragController) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+
+	b := ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedControlPlane{}).
+		WithOptions(controller.Options{
+			RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 10*time.Second),
+		}).Named(r.ControllerName)
+
+	r.log.Info("NewControllerManagedBy() called")
+	if _, err := b.Build(r); err != nil {
+		return err
+	}
+
+	endpointsFunc := func() ([]string, error) {
+		return r.etcdEndpoints(ctx)
+	}
+	r.etcdClient = etcdcli.NewEtcdClient(endpointsFunc, events.NewLoggingEventRecorder(r.ControllerName))
+
+	// Set this so that it will immediately requeue itself.
+	r.requeueTime = time.Now().Add(controllerRequeueDuration)
+	r.defragWaitDuration = minDefragWaitDuration
+
+	r.log.Info("Running mgr.Start()")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		r.log.Error(err, "problem running manager")
+		os.Exit(1)
+	}
+	return nil
+}
+
+func (r *DefragController) UpdateRequeueTime() {
+	r.requeueTime = time.Now().Add(controllerRequeueDuration)
+	r.log.Info("Periodic check complete, requeueing", "duration", controllerRequeueDuration, "requeue time", r.requeueTime)
+}
+
+func (r *DefragController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	// If we are before our requeue should happen
+	// then we want to return without requeue as this is not the timed Reconcile().
+	if time.Now().Before(r.requeueTime) {
+		// This Reconcile() is about something changing, ignore for now.
+		return ctrl.Result{}, nil
+	}
+
+	r.log.Info("Running periodic check for defrag", "time", time.Now(), "requeue time", r.requeueTime)
+
+	result := ctrl.Result{RequeueAfter: controllerRequeueDuration}
+
+	// Fetch the HostedControlPlane
+	hostedControlPlane := &hyperv1.HostedControlPlane{}
+	err := r.Client.Get(ctx, req.NamespacedName, hostedControlPlane)
+	if err != nil {
+		r.UpdateRequeueTime()
+		return result, err
+	}
+
+	// Return early if HostedControlPlane is deleted
+	if !hostedControlPlane.DeletionTimestamp.IsZero() {
+		r.UpdateRequeueTime()
+		return result, nil
+	}
+
+	// Run defragmentation.  This checks first to see if it needs it.
+	if err := r.runDefrag(ctx); err != nil {
+		r.UpdateRequeueTime()
```

Comment:
I think we want to retry as fast as possible on error, and not wait for the requeue duration.